pkg/services/ngalert/state/cache.go

```diff
@@ -209,19 +209,22 @@ func (c *cache) get(orgID int64, alertRuleUID, stateId string) *State {
 	return nil
 }
 
-func (c *cache) getAll(orgID int64) []*State {
+func (c *cache) getAll(orgID int64, skipNormalState bool) []*State {
```

Comment:
I wonder if we should dependency inject all of `FeatureToggles` rather than passing this bool around everywhere. I realize it's more overhead, but the bool hurts readability at call sites IMO.

And, if we decide we like this change and want to make it always-on (removing the toggle), it's going to be a pain to dig back through all this code and remove the args.

Reply:
Per my comment on https://github.com/grafana/grafana/pull/61201#discussion_r1066034793 I was hoping we could make an argument for removing the feature flag w/o necessarily incurring in a breaking change. 

I think the feature flag makes this more complicated and doesn't really buy us anything (it is a breaking change but a very subtle one) - could we maybe assume that until determined by the next evaluation when a rule starts it's normal in order to keep semantics equal? The way I understand this if an instance would have no state to recover from it's because it was either:

A) never present 
B) not recorded because it was normal

Reply:
Sure, that would solve my comment here too, I'm just not a fan of the boolean flags permeating the code that deeply change its behavior. If we do this a couple times, we end up with a large matrix of boolean toggles and things become hard to follow. Just having one path is a solution here - but if we must have two, I'd rather it be flagged each place it's used.

Reply:
My intention is to keep the feature flag until Grafana 10 when we can reason about breaking changes. Although there is only one breaking change, in the presentation, I think it is still a breaking one and should be handled accordingly - we should make it default in the next major version. 

re features toggle, I am really on the fence. I did that initially but then decided to use the flag because backtesting creates a state manager and therefore it needs a feature manager just to read that toggle, and create a fake one was kinda messy, as well as propagating feature manager. Anyway, I see your point, and will update to use toggles.

Reply:
I strongly agree with Alex's remark about using a boolean parameter to change the behavior of a function. That's a low-hanging fruit and we should change it. Have two functions or use the options pattern.

Reply:
If we abstract from the fact that this is a feature flag, the argument here makes perfect sense: a caller requests all abnormal states. I can agree that it could look obscure and unobvious in a public API like the constructor method of the state manager. However, I do not see any problems with methods, especially when it is the only boolean argument.
Providing an entire feature manager is what is very ambiguous and obscure because it makes it unclear what flags and how they are used. It is like providing an entire Grafana config just to read one field two layers down.

Reply:
> Providing an entire feature manager is what is very ambiguous

I'd agree with this for a lot of objects, but generally a feature flag structure intentionally reverses this rule:
1. It's clear at each point the flag is checked that it's a flagged code, not a permanent config toggle.
2. It makes the flag checks less intrusive, and easier to remove in the future.

Especially passing the boolean as part of the function signature, this will have to go through back and be refactored out later when we decide to drop the toggle and make this default behavior.

Reply:
I still think that the bool argument is better than the feature manager in this particular case because the cache implementation has different methods like `getOrCreate`, `removeByRuleUID`, `get` that return state(s). Conceptually, having a feature flag changing the behavior of just a subset of methods would make it less clear what is going on and why different methods of the cache return different data. 
the flag there is purely for optimization purposes - to avoid filtering the slice in the manager.